src/ESLint.Formatter/sarif-with-rules.js

```diff
@@ -0,0 +1,145 @@
+/**
+ * @fileoverview SARIF v2.0 formatter
+ * @author Chris Meyer
+ */
+
+"use strict";
+
+//------------------------------------------------------------------------------
+// Helper Functions
+//------------------------------------------------------------------------------
+
+/**
+ * Returns the severity of warning or error
+ * @param {Object} message message object to examine
+ * @returns {string} severity level
+ * @private
+ */
+function getResultLevel(message) {
+    if (message.fatal || message.severity === 2) {
+        return "error";
+    }
+    return "warning";
+}
+
+
+//------------------------------------------------------------------------------
+// Public Interface
+//------------------------------------------------------------------------------
+
+module.exports = function(results, rules) {
+    const sarifLog = {
+        version: "2.0.0",
+        $schema: "http://json.schemastore.org/sarif-2.0.0",
+        runs: [
+            {
+                tool: {
+                    name: "ESLint"
+                }
+            }
+        ]
+    };
+
+    const sarifFiles = {};
+    const sarifRules = {};
+    const sarifResults = [];
+
+    results.forEach(result => {
+
+        if (typeof sarifFiles[result.filePath] === "undefined") {
+
+            // Create a new entry in the files dictionary.
+            sarifFiles[result.filePath] = {
```

Comment:
>filePath [](start = 30, length = 8)

note that this code will need updating very soon due to converting files table to an array (no longer a dictionary)